usb: cdc_acm: TX memory and performance improvements. Avoid ZLP transfer

aurel32 opened the pull request:
Currently the cdc_acm implementation send data to the USB stack by
chunks of USB MPS. This has 2 drawbacks:
- when more data than USB MPS has to be sent, each chunk needs 2 packets
  due to the need of an extra ZLP;
- a temporary buffer of size USB MPS is needed.

This patch improves the memory consumption and performances by passing
the ring buffer directly to the USB stack. It only has a small
performance degradation when the ring buffer wraps and less than USB MPS
has to be sent.

Fixes #21713.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: cdc_acm: TX memory and performance improvements

Currently the cdc_acm implementation pass the data to usb_transfer() by
a chunk of IN endpoint MPS. This has 2 drawbacks:
- at higher throughput, each transfer needs 2 packets due to the need of
  an extra ZLP;
- a temporary buffer of size USB MPS is needed.

This patch improves the memory consumption and performances by passing
the ring buffer directly to usb_transfer(). It only has a small
performance degradation when the ring buffer wraps and less than a IN
endpoint MPS has to be sent.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

jfischer-phytec-iot commented:
> Currently the cdc_acm implementation send data to the USB stack by
> chunks of USB MPS. This has 2 drawbacks:
> 
>     * when more data than USB MPS has to be sent, each chunk needs 2 packets
>       due to the need of an extra ZLP;
> 
>     *

This not precise, usb_transfer adds ZLP when transfer length is multiple or equal endpoint MPS.

aurel32 commented:
> > Currently the cdc_acm implementation send data to the USB stack by
> > chunks of USB MPS. This has 2 drawbacks:
> > ```
> > * when more data than USB MPS has to be sent, each chunk needs 2 packets
> >   due to the need of an extra ZLP;
> > 
> > *
> > ```
> 
> This not precise, usb_transfer adds ZLP when transfer length is multiple or equal endpoint MPS.

Yes, in that case the USB CDC ACM code makes sure that what is passed to usb_transfer is less or equal to endpoint MPS. That's why an extra ZLP is added.

jfischer-no reviewed (changes requested):
Thanks. I suggest to tweak the commit message to point to the issue with temporary tx_buf.

```
At higher throughput, the cdc_acm implementation
pass the data to usb_transfer() by a chunk of IN endpoint MPS.
This has 2 drawbacks:
- each transfer needs 2 packets due to the need of an extra ZLP
- a temporary buffer of size USB MPS is needed.
```

jfischer-phytec-iot commented:
@carlescufi @emob-nordic That also allow to remove transfer of ZLP by setting CONFIG_USB_CDC_ACM_RINGBUF_SIZE lower than MPS of the IN endpoint. @aurel32 You could also add `Resolves: #21713` to the commit message.

carlescufi commented:
@emob-nordic and @jfischer-phytec-iot have you tested this on multiple platforms?

jfischer-phytec-iot commented:
> @emob-nordic and @jfischer-phytec-iot have you tested this on multiple platforms?

No, but that should not depend on a platform. I will do some test on hardware I have.

aurel32 commented:
> @carlescufi @emob-nordic That also allow to remove transfer of ZLP by setting CONFIG_USB_CDC_ACM_RINGBUF_SIZE lower than MPS of the IN endpoint (#21713 issue). 

I am not sure it's a good idea given that CONFIG_USB_CDC_ACM_RINGBUF_SIZE is the shared between both OUT and IN endpoints, and the same way CONFIG_CDC_ACM_BULK_EP_MPS is shared between both. If the ring buffer is smaller than MPS, and if the host send packets bigger than the ring buffer, they'll get partially dropped before the application has the time to consume them.

jfischer-phytec-iot commented:
> > @carlescufi @emob-nordic That also allow to remove transfer of ZLP by setting CONFIG_USB_CDC_ACM_RINGBUF_SIZE lower than MPS of the IN endpoint (#21713 issue).
> 
> I am not sure it's a good idea given that CONFIG_USB_CDC_ACM_RINGBUF_SIZE is the shared between both OUT and IN endpoints, and the same way CONFIG_CDC_ACM_BULK_EP_MPS is shared between both. If the ring buffer is smaller than MPS, and if the host send packets bigger than the ring buffer, they'll get partially dropped before the application has the time to consume them.

Right, missed that. We could split it in CONFIG_USB_CDC_ACM_RX_RINGBUF_SIZE and CONFIG_USB_CDC_ACM_TX_RINGBUF_SIZE.
CONFIG_CDC_ACM_BULK_EP_MPS or other *_EP_MPS is an other issue that should be cleaned up because for e.g. bulk EP it is either FS and 64 bytes or HS and 512 bytes. I started to look at the high speed but the driver fix for my test device is for two months in review.

aurel32 committed:
usb: cdc_acm: do not send ZLP for IN transactions

When the length of the transfer is an exact multiple of IN endpoint MPS
in the ring buffer, transfer one byte less to avoid zero-length packet.
Otherwise the application running on the host may conclude that there
is no more data to be received (i.e. the transaction has completed),
hence not triggering another I/O Request Packet (IRP).

Fixes #21713.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
> @carlescufi @emob-nordic That also allow to remove transfer of ZLP by setting CONFIG_USB_CDC_ACM_RINGBUF_SIZE lower than MPS of the IN endpoint (#21713 issue). @aurel32 You could also add `Resolves: #21713` to the commit message?

As discussed this is not a working solution. Instead I have just added an additional commit that reduces the length of the transfer by one byte in case in case there is an exact multiple of IN endpoint MPS in the ring buffer. This has no impact on the performances as this is still able to transfer multiple chunks of IN endpoint MPS, the difference being that the two last packets are of size (MPS-1, 1) instead of (MPS, 0). It's even possible that during the USB transfer the ring buffer is filled with additional bytes.

emob-nordic commented:
I will have a deeper look today and perform some tests.

jfischer-phytec-iot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -301,6 +301,18 @@ static void tx_work_handler(struct k_work *work)
+	 * IN transaction, and thus preventing any more data to be sent from
+	 * Zephyr to the Host.
+	 */
+	if (!(len % CONFIG_CDC_ACM_BULK_EP_MPS)) {
```
I think `len` should be checked for zero.

aurel32 replied:
I don't think this could actually happen, but I agree it's better to be safe than sorry. However should we just return from this function if it happens? From what I understand calling usb_transfer() with a length of 0 will send a ZLP, which is what we want to avoid. Note that this problem is not introduced by this PR, it was there also in the original code.

jfischer-phytec-iot replied:
I think it should just return if len == 0. Agree, the issue was already there but you patch escalate it a little :-)

jfischer-phytec-iot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -301,6 +301,18 @@ static void tx_work_handler(struct k_work *work)
 				 CONFIG_USB_CDC_ACM_RINGBUF_SIZE);
 
+	/*
+	 * Send less data to avoid having to send a zero-length packet. When
```
I suggest to avoid word "send" here, one could believe that it is possible to send data to the host like over Ethernet or UART.

~~Send less data to avoid having to send a zero-length packet. When
send a zero-length packet, the host or t~~he application running on
the host may believe that there is no more data to be received
(i.e. the transaction has completed), hence not triggering another
~~IN transaction, and thus preventing any more data to be sent from
Zephyr to the Host.~~

Transfer less data to avoid zero-length packet. The application running on
the host may conclude that there is no more data to be received
(i.e. the transaction has completed), hence not triggering another
I/O Request Packet (IRP).

aurel32 committed:
usb: cdc_acm: do not call usb_transfer if there is nothing to transfer

In case the ring buffer is empty, ring_buf_get_claim() returns a zero
length. Exit the function in that case as calling usb_transfer() with a
zero length will send a ZLP.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
@jfischer-phytec-iot I have just pushed a new version which check for a zero length and update the comments following your suggestion.

All that said I wonder if we could also simply claim (CDC_ACM_BUFFER_SIZE - 1) bytes from the ring buffer to make the code simpler and avoid sending a ZLP. This is slightly less efficient but with a 64-byte MPS, that's at most a difference of 1.5%.

jfischer-phytec-iot commented:
> All that said I wonder if we could also simply claim (CDC_ACM_BUFFER_SIZE - 1) bytes from the ring buffer to make the code simpler and avoid sending a ZLP. This is slightly less efficient but with a 64-byte MPS, that's at most a difference of 1.5%.

I am against that, there is no voodoo, the code is simple enough.
Thanks for the patches.

eobalski reviewed (approved):
Works fine for me.

carlescufi commented:
@aurel32 this is no longer a bugfix so it will need to wait until the 2.3 merge window is open.

jfischer-phytec-iot commented:
> @aurel32 this is no longer a bugfix so it will need to wait until the 2.3 merge window is open.

@jhedberg  @carlescufi This resolves the issue #21713 and I would like to see it in 2.2

jhedberg commented:
> @jhedberg @carlescufi This resolves the issue #21713 and I would like to see it in 2.2

We'd need to add the TSC label to it and add it to the "Pending TSC approval" column of https://github.com/zephyrproject-rtos/zephyr/projects/27 in that case. It will then be voted upon in today's TSC call.

nvlsianpu reviewed (approved):
Tested along with mcuboot recovery DFU via usb cdc acm which was affected (by https://github.com/zephyrproject-rtos/zephyr/issues/21713). Patch fixes the issue I had.

jfischer-phytec-iot commented:
> Tested along with mcuboot recovery DFU via usb cdc acm which was affected (by #21713). Patch fixes the issue I had.

@nvlsianpu could you report this issue in #21713 and change the label back to a bug, please?

finikorg reviewed (approved):
LGTM

carlescufi commented:
@jhedberg turns out this actually fixes a real bug in the CDC ACM class. Please merge as part of 2.2.
